apps/src/lab2/views/ExtraLinksModal.tsx

```diff
@@ -83,6 +102,144 @@ const ExtraLinksModal: React.FunctionComponent<ExtraLinksModalProps> = ({
     }
   };
 
+  const onBookmark = async () => {
+    try {
+      await HttpClient.put(
+        `/featured_projects/${channelId}/bookmark`,
+        undefined,
+        true,
+        {contentType: 'application/json;charset=UTF-8'}
+      );
+      setFeaturedProjectStatus(FeaturedProjectStatus.bookmarked);
+    } catch (e) {
+      console.log('Error bookmarking project', e);
+    }
+  };
+
+  const renderCloneLevel = (
+    <div>
+      <Button
+        size={Button.ButtonSize.small}
+        color={Button.ButtonColor.purple}
+        onClick={() => setShowCloneField(!showCloneField)}
+        text={showCloneField ? 'Cancel Clone' : 'Clone'}
+      />
+      {showCloneField && (
+        <div>
+          {'New level name: '}
+          <input
+            type="text"
+            value={clonedLevelName}
+            onChange={event => setClonedLevelName(event.target.value)}
+          />
+          <Button
+            onClick={handleClone}
+            text={'Clone'}
+            size={Button.ButtonSize.small}
+          />
+          {cloneError && (
+            <p className={moduleStyles.errorMessage}>{cloneError}</p>
+          )}
+        </div>
+      )}
+    </div>
+  );
+
+  const renderDeleteLevel = (
+    <div>
+      <Button
+        size={Button.ButtonSize.small}
+        text={showDeleteConfirm ? 'Cancel Delete' : 'Delete'}
+        color={Button.ButtonColor.red}
+        onClick={() => setShowDeleteConfirm(!showDeleteConfirm)}
+      />
+      {showDeleteConfirm && (
+        <div>
+          {'Are you sure you want to delete this level? '}
+          <Button
+            onClick={handleDelete}
+            text={'Confirm Delete'}
+            size={Button.ButtonSize.small}
+          />
+          {deleteError && (
+            <p className={moduleStyles.errorMessage}>{deleteError}</p>
+          )}
+        </div>
+      )}
+    </div>
+  );
+
+  const renderScriptLevelPathLinks = (
+    <>
+      <StrongText>
+        This level is in{' '}
+        {Object.entries(levelLinkData.script_level_path_links).length} scripts:
+      </StrongText>
+      <ul>
+        {levelLinkData.script_level_path_links.map((link, index) => (
+          <li key={index}>
+            <a href={'/s/' + link.script}>{link.script}</a> as{' '}
+            <a href={link.path}>{link.path}</a>
+          </li>
+        ))}
+      </ul>
+    </>
+  );
+
+  const renderRemixAncestry = (remixList: string[]) => {
+    if (remixList.length === 0) {
+      return <li>Not a remix.</li>;
+    }
+    return remixList.map((link: string, index: number) => (
+      <li key={index}>
+        <a href={link}>{link}</a>
+      </li>
+    ));
```

Comment:
This is probably okay, but it might be better to key off of `link` itself. That way we don't even need the `index` arg, and we're safe against render errors if the list is re-ordered.